rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -0,0 +1,390 @@
+   If input is dense/sparse, then output is dense/sparse."""
+
+  def __init__(self, max_tokens=None, num_oov_tokens=1, vocabulary=None,
+               name=None, **kwargs):
+    """Constructs a CategoryLookup layer.
+
+    Args:
+      max_tokens: The maximum size of the vocabulary for this layer. If None,
+              there is no cap on the size of the vocabulary. This is used when `adapt`
+              is called.
+      num_oov_tokens: Non-negative integer. The number of out-of-vocab tokens. 
+              All out-of-vocab inputs will be assigned IDs in the range of 
+              [0, num_oov_tokens) based on a hash. When
+              `vocabulary` is None, it will convert inputs in [0, num_oov_tokens)
+      vocabulary: the vocabulary to lookup the input. If it is a file, it represents the 
+              source vocab file; If it is a list/tuple, it represents the source vocab 
+              list; If it is None, the vocabulary can later be set.
+      name: Name to give to the layer.
+     **kwargs: Keyword arguments to construct a layer.
+
+    Input: a string or int tensor of shape `[batch_size, d1, ..., dm]`
+    Output: an int tensor of shape `[batch_size, d1, ..., dm]`
+
+    Example:
+
+    If one input sample is `["a", "c", "d", "a", "x"]` and the vocabulary is ["a", "b", "c", "d"],
+    and a single OOV token is used (`num_oov_tokens=1`), then the corresponding output sample is
+    `[1, 3, 4, 1, 0]`. 0 stands for an OOV token.
+    """
+    pass
+
+`tf.keras.layers.CategoryCrossing`
+CategoryCrossing(PreprocessingLayer):
+"""This layer transforms multiple categorical inputs to categorical outputs
+   by Cartesian product. and hash the output if necessary.
+   If any input is sparse, then output is sparse, otherwise dense."""
+
+  def __init__(self, depth=None, num_bins=None, name=None, **kwargs):
+    """Constructs a CategoryCrossing layer.
+    Args:
+      depth: depth of input crossing. By default None, all inputs are crossed
+             into one output. It can be an int or tuple/list of ints, where inputs are
+             combined into all combinations of output with degree of `depth`. For example,
+             with inputs `a`, `b` and `c`, `depth=2` means the output will be [ab;ac;bc]
+      num_bins: Number of hash bins. By default None, no hashing is performed.
+      name: Name to give to the layer.
+      **kwargs: Keyword arguments to construct a layer.
+
+    Input: a list of int tensors of shape `[batch_size, d1, ..., dm]`
+    Output: a single int tensor of shape `[batch_size, d1, ..., dm]`
+
+    Example:
+
+    If the layer receives two inputs, `a=[[1, 2]]` and `b=[[1, 3]]`,
+    and if depth is 2, then
+    the output will be a single integer tensor `[[i, j, k, l]]`, where:
+    i is the index of the category "a1=1 and b1=1"
+    j is the index of the category "a1=1 and b2=3"
+    k is the index of the category "a2=2 and b1=1"
+    l is the index of the category "a2=2 and b2=3"
```

Comment:
I don't understand this example. What is `a1` vs `a2`? What will the "single integer tensor `[[i, j, k, l]]`" actually look like for the given inputs?

Reply:
Yeah it is confusing. Updated.